### ISSUE TITLE
Errors being caused by allowing `conda-forge` & `dask/label/dev` packages to co-exist

### ISSUE THREAD

**Issue by galipremsagar** [dask/dask#9367]
**What happened**:

Looks like the way currently `dask`, `dask-core` & `distributed` are being packaged in `conda-forge` & `dask/label/dev` to be able to use interchangeably is leading to issues when being used with `dask-cuda-worker`.

**What you expected to happen**:
Ideally, I'd expect `dask/label/dev` packages to not be compatible with `conda-forge` packages, and vice-versa.

**Minimal Complete Verifiable Example**:
[sample_error.yml.txt](https://github.com/dask/dask/files/9286297/sample_error.yml.txt)

Error scenario:
```bash
(myenv) pgali@dt07:/nvme/0/pgali/cudf$ mamba env create --prefix /nvme/0/pgali/envs/cudfdev --file sample_error.yml
(myenv) pgali@dt07:/nvme/0/pgali/cudf$ conda activate /nvme/0/pgali/envs/cudfdev
(cudfdev) pgali@dt07:/nvme/0/pgali/cudf$ conda list | grep "dask"
dask                      2022.7.1           pyhd8ed1ab_0    conda-forge
dask-core                 2022.7.1a220722 py_gedfa91e22_16    dask/label/dev
dask-cuda                 22.08.00a220808 py39_g9f46a1a_35    rapidsai-nightly
(cudfdev) pgali@dt07:/nvme/0/pgali/cudf$ conda list | grep "distributed"
distributed               2022.7.1           pyhd8ed1ab_0    conda-forge

(cudfdev) pgali@dt07:/nvme/0/pgali/cudf$ ipython
iPython 3.9.13 | packaged by conda-forge | (main, May 27 2022, 16:56:21) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.4.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import dask

In [2]: dask.__version__
Out[2]: '2022.7.0+16.gedfa91e22'


(cudfdev) pgali@dt07:/nvme/0/pgali/cudf$ dask-cuda-worker
Traceback (most recent call last):
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/pkg_resources/__init__.py", line 581, in _build_master
    ws.require(__requires__)
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/pkg_resources/__init__.py", line 909, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/pkg_resources/__init__.py", line 800, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (dask 2022.7.0+16.gedfa91e22 (/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages), Requirement.parse('dask==2022.7.1'), {'dask-cuda', 'distributed'})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/nvme/0/pgali/envs/cudfdev/bin/dask-cuda-worker", line 33, in <module>
    sys.exit(load_entry_point('dask-cuda==22.8.0a0', 'console_scripts', 'dask-cuda-worker')())
  File "/nvme/0/pgali/envs/cudfdev/bin/dask-cuda-worker", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/importlib/metadata.py", line 86, in load
    module = import_module(match.group('module'))
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 972, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 972, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/dask_cuda/__init__.py", line 12, in <module>
    from .cuda_worker import CUDAWorker
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/dask_cuda/cuda_worker.py", line 22, in <module>
    from .initialize import initialize
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/dask_cuda/initialize.py", line 5, in <module>
    import numba.cuda
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/numba/__init__.py", line 19, in <module>
    from numba.core import config
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/numba/core/config.py", line 15, in <module>
    import llvmlite.binding as ll
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/llvmlite/binding/__init__.py", line 4, in <module>
    from .dylib import *
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/llvmlite/binding/dylib.py", line 3, in <module>
    from llvmlite.binding import ffi
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/llvmlite/binding/ffi.py", line 175, in <module>
    from pkg_resources import resource_filename
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/pkg_resources/__init__.py", line 3260, in <module>
    def _initialize_master_working_set():
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/pkg_resources/__init__.py", line 3234, in _call_aside
    f(*args, **kwargs)
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/pkg_resources/__init__.py", line 3272, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/pkg_resources/__init__.py", line 583, in _build_master
    return cls._build_from_requirements(__requires__)
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/pkg_resources/__init__.py", line 596, in _build_from_requirements
    dists = ws.resolve(reqs, Environment())
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/pkg_resources/__init__.py", line 795, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'dask==2022.7.1' distribution was not found and is required by dask-cuda, distributed

```

Working scenario when `dask/label/dev` is not present in the environment:
[sample_working.yml.txt](https://github.com/dask/dask/files/9286308/sample_working.yml.txt)

```bash
(myenv) pgali@dt07:/nvme/0/pgali/cudf$ mamba env create --prefix /nvme/0/pgali/envs/cudfdev --file sample_working.yml
(myenv) pgali@dt07:/nvme/0/pgali/cudf$ conda activate /nvme/0/pgali/envs/cudfdev
(cudfdev) pgali@dt07:/nvme/0/pgali/cudf$ dask-cuda-worker
# Works fine without any errors
(cudfdev) pgali@dt07:/nvme/0/pgali/cudf$ conda list | grep "dask"
dask                      2022.7.1           pyhd8ed1ab_0    conda-forge
dask-core                 2022.7.1           pyhd8ed1ab_0    conda-forge
dask-cuda                 22.08.00a220808 py39_g9f46a1a_35    rapidsai-nightly
(cudfdev) pgali@dt07:/nvme/0/pgali/cudf$ conda list | grep "distributed"
distributed               2022.7.1           pyhd8ed1ab_0    conda-forge

(cudfdev) pgali@dt07:/nvme/0/pgali/cudf$ ipython
Type 'copyright', 'credits' or 'license' for more information
IPython 8.4.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import dask

In [2]: dask.__version__
Out[2]: '2022.7.1'

```

**Comment by jrbourbeau** [comment]
Thanks for surfacing @galipremsagar. Looking at those `conda list` outputs, the thing that stands out to me is (in the failing case) `dask` and `dask-core` are being installed from different channels -- `conda-forge` and `dask/label/dev`, respectively. Naively I would have expected those packages to be installed from the same channel, given the channel priorities being used. That said, I'm not sure what packages `dask-cuda` is bringing in. Maybe the nightly version of that package is leading to the channel mismatch? cc @jakirkham @charlesbluca @pentschev in case you have any thoughts

**Comment by charlesbluca** [comment]
Don't think this is related to the dask-cuda nightly, as the same issue comes up with a more minimal conda env specification:

```
name: cudf_dev
channels:
  - dask/label/dev
  - conda-forge
dependencies:
  - dask==2022.7.1
  - distributed==2022.7.1
```

Think what's happening here is that:

- we specified a strict pinning to `dask==2022.7.1`; this package isn't available in `dask/label/dev` (only packages with versions like `2022.7.1a.*`) so we end up pulling it from conda-forge
- the conda-forge `dask` package for `2022.7.1` loosely pins `dask-core` to  `2022.7.1` (i.e. `2022.7.1.*`) which allows for nightly `dask-core` packages; since `dask/label/dev` is higher priority than conda-forge in the environment file, conda ends up pulling from there and we have a mixed stable/nightly environment

I think the fix for the conda-forge `dask` package should be relatively easy; we're currently specifying the `dask-core` and `distributed` like

```
    - dask-core {{ version }}
    - distributed {{ version }}
```

Which IIUC is equivalent to loosely pinning to `{{ version }}`; to ensure that we are only able to install `{{ version }}` exactly, think we would want to do something like:

```
    - dask-core =={{ version }}
    - distributed =={{ version }}
```

For the nightly `dask` packages, things are a little more complicated; since they're only built on commits to Distributed, we had to intentionally leave the `dask-core` dependency loosely pinned so that we could pull in newer `dask-core` nightlies that were built afterwards, which has the unfortunate side effect of allowing nightly `dask` to be installed with conda-forge `dask-core` (the opposite of what's happening here); I think we can resolve this by modifying the `dask` nightly recipe to have an upper bound constraint that effectively forces it to use the nightly channel, thinking of something like

```
    - dask-core >=2022.7.1a220711,<2022.7.1
```

**Comment by jakirkham** [comment]
Thanks all for the engagement here! 🙏

One consideration with `dask` & `distributed` packages is we may want to allow hot-fixes (like `2022.8.1.1`) for one of the packages without making a full release. If that is something we want, then `==` may be too tight to allow for that update.

Another thing that we may want to look at is whether this is still an issue with [strict channel priority]( https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-channels.html#strict-channel-priority ). Would think in that case we get everything from `dask/label/dev` instead of getting some packages from `conda-forge`.

**Comment by jrbourbeau** [comment]
I believe this is closed by https://github.com/dask/distributed/pull/6862, https://github.com/conda-forge/dask-feedstock/pull/190, and https://github.com/conda-forge/distributed-feedstock/pull/217. @galipremsagar I'm going to close this issue, but feel free to re-open if you're still experiencing this issue
